File: src/remedy/connect/bind.py
"""Chosen-IPv4 bind helpers. Wildcard binds (0.0.0.0 / :: / *) are refused."""

from __future__ import annotations

import ipaddress
import socket

WILDCARD = frozenset({"0.0.0.0", "::", "[::]", "*", "0:0:0:0:0:0:0:0"})
_LIMITED_BROADCAST = ipaddress.IPv4Address("255.255.255.255")

# Virtual NAT ranges a phone on the LAN can never reach (WSL2, Docker,
# Hyper-V internal, CGNAT, link-local). Used only to demote an address when a
# real default-route source exists — never to reject an explicit user pick.
_VIRTUAL_NAT_NETS = (
    ipaddress.ip_network("172.16.0.0/12"),
    ipaddress.ip_network("100.64.0.0/10"),
    ipaddress.ip_network("169.254.0.0/16"),
)


def _ipv4_key(ip: str) -> tuple[int, int, int, int]:
    """Numeric sort key so 10.x < 172.x < 192.x like humans expect."""
    try:
        return tuple(int(part) for part in ip.split("."))  # type: ignore[return-value]
    except ValueError:
        return (0, 0, 0, 0)


def is_wildcard_bind(host: str) -> bool:
    text = (host or "").strip()
    if not text:
        return False
    if text in WILDCARD:
        return True
    inner = text[1:-1] if text.startswith("[") and text.endswith("]") else text
    if inner in WILDCARD:
        return True
    try:
        return ipaddress.ip_address(inner).is_unspecified
    except ValueError:
        return False


def assert_chosen_bind(host: str) -> str:
    """Raise ValueError if wildcard or empty. Return stripped host."""
    text = (host or "").strip()
    if not text:
        raise ValueError("bind host is empty")
    if is_wildcard_bind(text):
        raise ValueError("wildcard bind is not allowed")
    return text


def is_chosen_ipv4(host: str) -> bool:
    """True for a unicast IPv4 (including 127.0.0.1). False for wildcard/hostname/empty."""
    text = (host or "").strip()
    if not text or is_wildcard_bind(text):
        return False
    try:
        ip = ipaddress.IPv4Address(text)
    except ValueError:
        return False
    return not (ip.is_unspecified or ip.is_multicast or ip == _LIMITED_BROADCAST)


def is_loopback_ipv4(host: str) -> bool:
    text = (host or "").strip()
    try:
        return bool(ipaddress.IPv4Address(text).is_loopback)
    except ValueError:
        return False


def prefer_lan_ipv4(
    addrs: list[str] | tuple[str, ...] | set[str],
    preferred: list[str] | tuple[str, ...] | set[str] = (),
) -> list[str]:
    """LAN unicast first, loopback last. Drops non-chosen addresses.

    ``preferred`` addresses (the default-route source) sort before all other
    LAN unicast — the real NIC must beat virtual NAT adapters even when its
    octets sort higher. Ordering is numeric, never lexicographic.
    """
    lan: list[str] = []
    loop: list[str] = []
    seen: set[str] = set()
    for raw in addrs:
        ip = str(raw or "").strip()
        if not ip or ip in seen or not is_chosen_ipv4(ip):
            continue
        seen.add(ip)
        if is_loopback_ipv4(ip):
            loop.append(ip)
        else:
            lan.append(ip)
    pre = [str(p).strip() for p in preferred if str(p or "").strip() in seen]
    rest = [ip for ip in lan if ip not in pre]
    return sorted(pre, key=_ipv4_key) + sorted(rest, key=_ipv4_key) + sorted(loop, key=_ipv4_key)


def pick_default_ipv4(addrs: list[str] | tuple[str, ...] | set[str] | None = None) -> str:
    """First LAN unicast (default-route source wins). Loopback only when alone.

    No args: uses the route-first order from :func:`list_candidate_ipv4`.
    Explicit list: numeric sort still applies (10.x < 172.x < 192.x).
    """
    rows = (
        list_candidate_ipv4()
        if addrs is None
        else prefer_lan_ipv4(addrs)
    )
    for ip in rows:
        if not is_loopback_ipv4(ip):
            return ip
    return rows[0] if rows else ""


def default_route_ipv4() -> str:
    """Source IPv4 the OS would use to reach the internet (the real LAN NIC).

    Empty when no route is available. This is the address a phone on the same
    LAN can actually reach — virtual NAT adapters never win here.
    """
    for dest in (("8.8.8.8", 80), ("1.1.1.1", 80)):
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
                sock.connect(dest)
                ip = sock.getsockname()[0]
                if is_chosen_ipv4(str(ip)):
                    return str(ip)
        except OSError:
            continue
    return ""


def _is_virtual_nat_ipv4(host: str) -> bool:
    text = (host or "").strip()
    try:
        addr = ipaddress.IPv4Address(text)
    except ValueError:
        return False
    return any(addr in net for net in _VIRTUAL_NAT_NETS)


def reachable_lan_host(configured: str) -> str:
    """Best LAN host to advertise/bind for phone pairing.

    The default-route source is what a phone on the same LAN can reach. A
    configured host that is a virtual-NAT address (WSL/Docker/Hyper-V/CGNAT)
    is demoted in favor of the real route; an explicit non-virtual pick is
    kept. Loopback only when nothing else exists.
    """
    cfg = str(configured or "").strip()
    route = default_route_ipv4()
    if route and (not cfg or is_loopback_ipv4(cfg) or _is_virtual_nat_ipv4(cfg)):
        return route
    return cfg or pick_default_ipv4()


def list_candidate_ipv4() -> list[str]:
    """Host IPv4s excluding 0.0.0.0. Default-route source first; 127.0.0.1 last."""
    found: set[str] = {"127.0.0.1"}
    try:
        hostname = socket.gethostname()
        _, _, addrs = socket.gethostbyname_ex(hostname)
        found.update(ip for ip in addrs if is_chosen_ipv4(ip))
    except OSError:
        pass
    try:
        for info in socket.getaddrinfo(socket.gethostname(), None, socket.AF_INET):
            ip = info[4][0]
            if is_chosen_ipv4(str(ip)):
                found.add(str(ip))
    except OSError:
        pass
    route = default_route_ipv4()
    if route:
        found.add(route)
    return prefer_lan_ipv4(found, preferred=(route,) if route else ())


def is_global_ipv6(host: str) -> bool:
    """True for a global unicast IPv6 (not link-local, ULA, loopback, or wildcard)."""
    text = (host or "").strip().strip("[]").split("%", 1)[0]
    if not text or is_wildcard_bind(text):
        return False
    try:
        ip = ipaddress.IPv6Address(text)
    except ValueError:
        return False
    return bool(ip.is_global)


def list_candidate_ipv6() -> list[str]:
    """Global unicast IPv6 addresses on this host, if any."""
    found: set[str] = set()
    try:
        for info in socket.getaddrinfo(socket.gethostname(), None, socket.AF_INET6):
            raw = str(info[4][0]).split("%", 1)[0]
            if is_global_ipv6(raw):
                found.add(raw)
    except OSError:
        pass
    return sorted(found)


_TAILSCALE_NET = ipaddress.ip_network("100.64.0.0/10")


def tailscale_ipv4() -> str:
    """IPv4 on the Tailscale tailnet (100.64.0.0/10), or ``""``.

    The tailnet address is reachable from a paired phone on Wi-Fi *and* mobile
    data (Tailscale DERP relays do the NAT traversal), so when Tailscale is up
    this becomes the primary QR candidate and a second Connect listener binds
    it. Empty when Tailscale is not installed/connected.
    """
    for ip in list_candidate_ipv4():
        try:
            addr = ipaddress.IPv4Address(ip)
        except ValueError:
            continue
        if addr in _TAILSCALE_NET:
            return str(addr)
    return ""

File: native/go/connect/bind.go
package connect

import (
	"errors"
	"fmt"
	"net"
	"os"
	"sort"
	"strings"
)

// ErrBind is a Connect bind-policy failure (wildcard / empty / not chosen IPv4).
var ErrBind = errors.New("connect bind refused")

// WildcardHosts that must never be used for the Connect listener.
var WildcardHosts = map[string]struct{}{
	"0.0.0.0":         {},
	"::":              {},
	"[::]":            {},
	"*":               {},
	"0:0:0:0:0:0:0:0": {},
}

// Virtual NAT / CGNAT / link-local ranges a phone on the LAN cannot reach.
// Used only to demote an address when a real default-route source exists.
var virtualNATNets = []*net.IPNet{
	mustCIDR("172.16.0.0/12"),
	mustCIDR("100.64.0.0/10"),
	mustCIDR("169.254.0.0/16"),
}

var tailscaleNet = mustCIDR("100.64.0.0/10")

func mustCIDR(s string) *net.IPNet {
	_, n, err := net.ParseCIDR(s)
	if err != nil {
		panic(err)
	}
	return n
}

// IsWildcardBind reports whether host is an unspecified / catch-all bind.
func IsWildcardBind(host string) bool {
	text := strings.TrimSpace(host)
	if text == "" {
		return false
	}
	if _, ok := WildcardHosts[text]; ok {
		return true
	}
	inner := text
	if strings.HasPrefix(text, "[") && strings.HasSuffix(text, "]") {
		inner = text[1 : len(text)-1]
	}
	if _, ok := WildcardHosts[inner]; ok {
		return true
	}
	ip := net.ParseIP(inner)
	if ip == nil {
		return false
	}
	return ip.IsUnspecified()
}

// AssertChosenBind refuses empty and wildcard hosts. Returns the stripped host.
func AssertChosenBind(host string) (string, error) {
	text := strings.TrimSpace(host)
	if text == "" {
		return "", fmt.Errorf("%w: bind host is empty", ErrBind)
	}
	if IsWildcardBind(text) {
		return "", fmt.Errorf("%w: wildcard bind is not allowed", ErrBind)
	}
	return text, nil
}

// IsChosenIPv4 is true for a unicast IPv4 (including 127.0.0.1).
func IsChosenIPv4(host string) bool {
	text := strings.TrimSpace(host)
	if text == "" || IsWildcardBind(text) {
		return false
	}
	ip := net.ParseIP(text)
	if ip == nil {
		return false
	}
	v4 := ip.To4()
	if v4 == nil {
		return false
	}
	if ip.IsUnspecified() || ip.IsMulticast() {
		return false
	}
	if v4[0] == 255 && v4[1] == 255 && v4[2] == 255 && v4[3] == 255 {
		return false
	}
	return true
}

// IsLoopbackIPv4 reports whether host is a loopback IPv4.
func IsLoopbackIPv4(host string) bool {
	text := strings.TrimSpace(host)
	ip := net.ParseIP(text)
	if ip == nil {
		return false
	}
	v4 := ip.To4()
	if v4 == nil {
		return false
	}
	return ip.IsLoopback()
}

func ipv4Key(ip string) [4]int {
	parts := strings.Split(ip, ".")
	if len(parts) != 4 {
		return [4]int{}
	}
	var key [4]int
	for i, p := range parts {
		n := 0
		for _, c := range p {
			if c < '0' || c > '9' {
				return [4]int{}
			}
			n = n*10 + int(c-'0')
		}
		key[i] = n
	}
	return key
}

func sortIPv4(addrs []string) {
	sort.Slice(addrs, func(i, j int) bool {
		a, b := ipv4Key(addrs[i]), ipv4Key(addrs[j])
		for k := 0; k < 4; k++ {
			if a[k] != b[k] {
				return a[k] < b[k]
			}
		}
		return false
	})
}

// PreferLANIPv4 returns LAN unicast first (preferred route sources ahead of
// other LAN), then loopback. Drops non-chosen addresses.
func PreferLANIPv4(addrs []string, preferred ...string) []string {
	lan := make([]string, 0, len(addrs))
	loop := make([]string, 0, 1)
	seen := make(map[string]struct{}, len(addrs))
	for _, raw := range addrs {
		ip := strings.TrimSpace(raw)
		if ip == "" {
			continue
		}
		if _, ok := seen[ip]; ok {
			continue
		}
		if !IsChosenIPv4(ip) {
			continue
		}
		seen[ip] = struct{}{}
		if IsLoopbackIPv4(ip) {
			loop = append(loop, ip)
		} else {
			lan = append(lan, ip)
		}
	}
	pre := make([]string, 0, len(preferred))
	preSet := make(map[string]struct{}, len(preferred))
	for _, p := range preferred {
		ip := strings.TrimSpace(p)
		if ip == "" {
			continue
		}
		if _, ok := seen[ip]; !ok {
			continue
		}
		if _, ok := preSet[ip]; ok {
			continue
		}
		preSet[ip] = struct{}{}
		pre = append(pre, ip)
	}
	rest := make([]string, 0, len(lan))
	for _, ip := range lan {
		if _, ok := preSet[ip]; !ok {
			rest = append(rest, ip)
		}
	}
	sortIPv4(pre)
	sortIPv4(rest)
	sortIPv4(loop)
	out := make([]string, 0, len(pre)+len(rest)+len(loop))
	out = append(out, pre...)
	out = append(out, rest...)
	out = append(out, loop...)
	return out
}

// DefaultRouteIPv4 is the source IPv4 the OS would use to reach the internet.
// Empty when no route is available.
func DefaultRouteIPv4() string {
	for _, dest := range []string{"8.8.8.8:80", "1.1.1.1:80"} {
		conn, err := net.Dial("udp", dest)
		if err != nil {
			continue
		}
		local, ok := conn.LocalAddr().(*net.UDPAddr)
		_ = conn.Close()
		if !ok || local == nil || local.IP == nil {
			continue
		}
		ip := local.IP.To4()
		if ip == nil {
			continue
		}
		text := ip.String()
		if IsChosenIPv4(text) {
			return text
		}
	}
	return ""
}

func isVirtualNATIPv4(host string) bool {
	ip := net.ParseIP(strings.TrimSpace(host))
	if ip == nil {
		return false
	}
	v4 := ip.To4()
	if v4 == nil {
		return false
	}
	for _, n := range virtualNATNets {
		if n.Contains(v4) {
			return true
		}
	}
	return false
}

// PickDefaultIPv4 returns the first LAN unicast from addrs (or from
// ListCandidateIPv4 when addrs is nil). Loopback only when alone.
func PickDefaultIPv4(addrs []string) string {
	var rows []string
	if addrs == nil {
		rows = ListCandidateIPv4()
	} else {
		rows = PreferLANIPv4(addrs)
	}
	for _, ip := range rows {
		if !IsLoopbackIPv4(ip) {
			return ip
		}
	}
	if len(rows) > 0 {
		return rows[0]
	}
	return ""
}

// ReachableLANHost picks the best LAN host to advertise for phone pairing.
// Virtual-NAT / loopback / empty configured hosts are demoted to the default
// route when one exists; an explicit non-virtual pick is kept.
func ReachableLANHost(configured string) string {
	cfg := strings.TrimSpace(configured)
	route := DefaultRouteIPv4()
	if route != "" && (cfg == "" || IsLoopbackIPv4(cfg) || isVirtualNATIPv4(cfg)) {
		return route
	}
	if cfg != "" {
		return cfg
	}
	return PickDefaultIPv4(nil)
}

// ListCandidateIPv4 returns host IPv4s excluding wildcards. Default-route
// source first; 127.0.0.1 last.
func ListCandidateIPv4() []string {
	found := map[string]struct{}{"127.0.0.1": {}}
	hostname, err := osHostname()
	if err == nil && hostname != "" {
		if addrs, err := net.LookupHost(hostname); err == nil {
			for _, ip := range addrs {
				if IsChosenIPv4(ip) {
					found[ip] = struct{}{}
				}
			}
		}
		if infos, err := net.LookupIP(hostname); err == nil {
			for _, ip := range infos {
				v4 := ip.To4()
				if v4 == nil {
					continue
				}
				text := v4.String()
				if IsChosenIPv4(text) {
					found[text] = struct{}{}
				}
			}
		}
	}
	route := DefaultRouteIPv4()
	if route != "" {
		found[route] = struct{}{}
	}
	addrs := make([]string, 0, len(found))
	for ip := range found {
		addrs = append(addrs, ip)
	}
	if route != "" {
		return PreferLANIPv4(addrs, route)
	}
	return PreferLANIPv4(addrs)
}

// IsGlobalIPv6 is true for a global unicast IPv6 (not link-local, ULA,
// loopback, or wildcard).
func IsGlobalIPv6(host string) bool {
	text := strings.TrimSpace(host)
	text = strings.Trim(text, "[]")
	if i := strings.IndexByte(text, '%'); i >= 0 {
		text = text[:i]
	}
	if text == "" || IsWildcardBind(text) {
		return false
	}
	ip := net.ParseIP(text)
	if ip == nil || ip.To4() != nil {
		return false
	}
	return ip.IsGlobalUnicast()
}

// ListCandidateIPv6 returns global unicast IPv6 addresses on this host.
func ListCandidateIPv6() []string {
	found := map[string]struct{}{}
	hostname, err := osHostname()
	if err != nil || hostname == "" {
		return nil
	}
	infos, err := net.LookupIP(hostname)
	if err != nil {
		return nil
	}
	for _, ip := range infos {
		if ip.To4() != nil {
			continue
		}
		text := ip.String()
		if i := strings.IndexByte(text, '%'); i >= 0 {
			text = text[:i]
		}
		if IsGlobalIPv6(text) {
			found[text] = struct{}{}
		}
	}
	out := make([]string, 0, len(found))
	for ip := range found {
		out = append(out, ip)
	}
	sort.Strings(out)
	return out
}

// TailscaleIPv4 returns an IPv4 on the Tailscale CGNAT range, or "".
func TailscaleIPv4() string {
	for _, ip := range ListCandidateIPv4() {
		parsed := net.ParseIP(ip)
		if parsed == nil {
			continue
		}
		v4 := parsed.To4()
		if v4 != nil && tailscaleNet.Contains(v4) {
			return ip
		}
	}
	return ""
}

// osHostname is overridable in tests.
var osHostname = os.Hostname
